doc/manual/contract.md

```diff
@@ -0,0 +1,505 @@
+
+Within an `ensure` contract we can refer to the return value of the method with the keyword `result`. In the `Stack` class, the `[]` method guarantees that, the return value is the same as the element stored in the stack at the given index.
+
+~~~
+fun [](index: Int): T is ensure(result == data[index])
+~~~
+
+Note that the keyword `result` can only be used when the method has a return parameter.
+
+
+In `ensure` we can refer to any arguments and instance variables before the method call with the using of `old` keyword. Again, with the `pop` method, we can ensure that the `size` of the stack after the method call will be the same as before minus one `old(size) - 1 == size`.
+
+~~~
+fun pop(object: T) is ensure(old(size) - 1 == size)
+~~~
+
+In an `old(expression)` the `expression` is evaluated in the same context as the called method. Thus, it's possible to refer to attributes, methods, and arguments.
+
+`old(object)` stores a reference to object. By doing so, if `object` is mutated in the method the `old(object)` will also be mutated.
+
+Let us take a look on the following example:
+
+~~~
+class Boat
+	var position: Position
+
+	fun move(direction: Position)
+	do
+		position =  position + direction
+	end
+end
+
+class Position
+	var x: Float
+	var y: Float
+
+
+	fun +(position: Position): Position
+	is
+		ensure(old(x) + position.x == x)
+		nit ensure(old(y) +  position.y == y)
+	do
+		_x += position.x
+		_y += position.y
+		return self
+	end
+
+	redef fun ==(other)
+	is
+		ensure(result == (_x == other.x and _y == other.y))
+	do
+		if not other isa Position then return false
+		return _x == other.x and _y == other.y
+	end
+end
+~~~
+
+This example introduce the following problem: How to define the postcondition of the `move` method to indicate that the new position is equal to the old position plus direction?
+
+As we store a reference it's impossible to proceed like this `ensure(old(position) + direction == position)`. Several solutions exist to solve this problem:
+
+- You can capture the changes individually like as the `+` method in the `Position` class. Which would give the following postcondition for the `move` method:
+
+~~~
+fun move(direction: Position) is ensure( old(position.x) + direction.x  == position.x and old(position.y) + direction.y  == position.y)
+~~~
+
+- You can duplicate the object `position` with a method like `clone`, so that `not old(position.clone).is_same_instance(position)` is true.
+
+~~~
+class Position
+	super Cloneable
+	var x: Float
+	var y: Float
+
+	redef fun clone: SELF do return new Position(x,y)
+end
+
+class Boat
+	var position: Position
+
+	fun move(direction: Position)
+	is
+		ensure(old(position.clone) + direction  == position)
+	do
+		position =  position + direction
+	end
+end
+~~~
+
+Note that it is often better to capture only the interesting part, rather than providing a duplication method. The duplication method is generally reserved when you deal with complex objects.
+
+## Writing invariants
```

Comment:
Since invariants are at the top of your example and are also easier to explain, maybe you should start by them instead of the pre/post conditions?